kv/memberlist/memberlist_client.go

```diff
@@ -533,43 +529,83 @@ func (m *KV) JoinMembers(members []string) (int, error) {
 	return m.memberlist.Join(members)
 }
 
-func (m *KV) joinMembersOnStartup(ctx context.Context, members []string) error {
-	reached, err := m.memberlist.Join(members)
-	if err == nil {
-		level.Info(m.logger).Log("msg", "joined memberlist cluster", "reached_nodes", reached)
-		return nil
+// fastJoinMembersOnStartup attempts to reach small subset of nodes (computed as RetransmitMult * log10(number of discovered members + 1)).
+func (m *KV) fastJoinMembersOnStartup(ctx context.Context) {
+	startTime := time.Now()
+
+	members := m.discoverMembers(ctx, m.cfg.JoinMembers)
+	toReach := m.cfg.RetransmitMult * int(math.Ceil(math.Log10(float64(len(members)+1))))
+	if toReach > len(members) {
+		toReach = len(members)
 	}
 
-	if m.cfg.MaxJoinRetries <= 0 {
-		return err
+	level.Info(m.logger).Log("msg", "memberlist fast-join starting", "members_found", len(members), "to_reach", toReach)
```

Comment:
We're calling this feature "joining cluster", so why not being consistent here?

```suggestion
	level.Info(m.logger).Log("msg", "memberlist fast-join starting", "members_found", len(members), "to_join", toReach)
```